pvanalytics/features/shading.py

```diff
@@ -0,0 +1,368 @@
+        limit_direction='both'
+    ).to_numpy()
+    # set night to nan
+    ghi_image[~_to_image(daytime.to_numpy(), image_width)] = np.nan
+    return (
+        ghi_image,
+        _to_image(scaled_clearsky.to_numpy(), image_width),
+        clouds_image,
+        image_times
+    )
+
+
+def _detect_clouds(ghi, clearsky_ghi, duration='50T'):
+    """Use :py:func:`pvanalytics.clearsky.reno` to detect clouds.
+
+    Returns
+    -------
+    Series
+        Boolean series with true for cloudy periods that are at least as long
+        as `duration`.
+    """
+    clouds = pvlib.clearsky.detect_clearsky(
+        ghi,
+        clearsky_ghi,
+        ghi.index,
+        window_length=10
+    )
+    return clouds.rolling(duration).sum() == 0
+
+
+def _remove_pillars(wires):
+    j = np.array([[0, 0, 0, 0, 0],
+                  [0, 0, 1, 0, 0],
+                  [0, 0, 1, 0, 0]])
+    k = np.array([[1, 0, 0, 0, 1],
+                  [1, 0, 0, 0, 1],
+                  [1, 0, 0, 0, 1]])
+    a1 = np.logical_and(
+        wires,
+        np.logical_not(ndimage.binary_hit_or_miss(wires, j, k))
+    )
+    a2 = np.logical_and(
+        wires,
+        np.logical_not(
+            ndimage.binary_hit_or_miss(wires, np.flipud(j), np.flipud(k))
+        )
+    )
+    return np.logical_and(a1, a2)
+
+
+def _fill_gaps(wires, out):
+    """Fill 1-minute gaps."""
+    mask = np.array([[1, 0, 1]])
+    return np.logical_or(
+        wires, ndimage.binary_hit_or_miss(wires, mask), out=out)
+
+
+def _remove_spikes(wires, out):
+    """Remove 1-day spikes."""
+    mask = np.array([[0, 1, 0]]).T
+    temp_image = np.ndarray(wires.shape)
+    hit_miss = ndimage.binary_hit_or_miss(wires, mask)
+    return np.logical_and(
+        wires,
+        np.logical_not(
+            hit_miss,
+            out=temp_image
+        ),
+        out=out
+    )
+
+
+def _restore_gaps(wires):
+    """Restore 2-minute and 1-minute gaps"""
+    wires = np.logical_or(
+        wires,
+        ndimage.binary_hit_or_miss(
+            wires,
+            np.array([[0, 1, 0, 0, 1]]), np.array([[0, 0, 1, 1, 0]])
+        )
+    )
+    return np.logical_or(
+        wires,
+        ndimage.binary_hit_or_miss(wires, np.array([[1, 0, 1]]))
+    )
+
+
+def _height(bbox):
+    """Return the height of a bounding box.
+
+    Parameters
+    ----------
+    bbox : tuple
+        4-tuple specifying ``(min_row, min_col, max_row, max_col)``.
+
+    Returns
+    -------
+    int
+        The height of the bounding box in pixels.
+    """
+    return bbox[2] - bbox[0] + 1
+
+
+def _filter_blobs(wires, blob_length, connectivity):
+    """Remove blobs shorter than blob_length.
+
+    Examines all connected components in the image and removes any
+    that span fewer than `blob_length` columns.
+    """
+    components = measure.label(wires, connectivity=connectivity, background=0)
+    keep = [props.label for props in measure.regionprops(components)
+            if _height(props.bbox) > blob_length]
```

Comment:
Here is one discrepancy from the MATLAB code:

The comments indicate that we should eliminate blobs that span fewer than `blob_length` days; however, in `GHIImage.bwlengthopen` it looks like the width of the bounding box is used instead of the height (`cc` is the connected component):

```matlab
                [~,cols] = ind2sub(size(cc),find(cc == ii));
                len(ii) = max(cols)-min(cols)+1;
```

I found that for the SMUD test data using width instead of height resulted in many clouds being left in the image.

Reply:
Matlab images had time of day on the row index, and day of year on the column index. Here we are transposed from that. This doesn't affect the blob detection (since it is already based on `_height()`, but did necessitate a change to the remove pillars step (transposing the masks).